test/util/malicious/hasher.go

```diff
@@ -3,28 +3,50 @@ package malicious
 import (
 	"bytes"
 	"errors"
+	"fmt"
 	"hash"
 
 	"github.com/celestiaorg/nmt"
 	"github.com/celestiaorg/nmt/namespace"
 )
 
 // NOTE: This file is a copy of the original nmt.Hasher implementation, but with
-// all of the validation logic removed. This allows for malicious nodes to create nmt roots that are out of order.
+// all of the validation logic removed. This allows for malicious nodes to
+// create nmt roots that are out of order. The documentation of this file may be
+// incorrect.
 
 const (
 	LeafPrefix = 0
 	NodePrefix = 1
 )
 
-var _ hash.Hash = (*Hasher)(nil)
+var _ hash.Hash = (*NmtHasher)(nil)
 
 var (
-	ErrInvalidNodeLen = errors.New("invalid NMT node size")
-	ErrInvalidLeafLen = errors.New("invalid NMT leaf size")
+	ErrUnorderedSiblings         = errors.New("NMT sibling nodes should be ordered lexicographically by namespace IDs")
+	ErrInvalidNodeLen            = errors.New("invalid NMT node size")
+	ErrInvalidLeafLen            = errors.New("invalid NMT leaf size")
+	ErrInvalidNodeNamespaceOrder = errors.New("invalid NMT node namespace order")
 )
 
-type Hasher struct {
+// Hasher describes the interface nmts use to hash leafs and nodes.
+//
+// Note: it is not advised to create alternative hashers if following the
+// specification is desired. The main reason this exists is to not follow the
+// specification for testing purposes.
+type Hasher interface {
+	IsMaxNamespaceIDIgnored() bool
+	NamespaceSize() namespace.IDSize
+	HashLeaf(data []byte) ([]byte, error)
+	HashNode(leftChild, rightChild []byte) ([]byte, error)
+	EmptyRoot() []byte
+}
+
+var _ Hasher = &NmtHasher{}
```

Comment:
went ahead and updated the hasher using the most recent nmt version. we might want to automate this in the future.

this time, I tried to keep the changes as minimal as possible. This means that instead of remove all the validation entirely, we're now only removing the pieces specific to ordering the namespaces